src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageSearch/PackageSearchRunner.cs

```diff
@@ -0,0 +1,257 @@
+
+            foreach (var packageSource in listEndpoints)
+            {
+                Task<IEnumerable<IPackageSearchMetadata>> searchTask = searchPackageSourceAsync(packageSource);
+                searchRequests.Add(searchTask, packageSource);
+            }
+
+            IPackageSearchResultRenderer packageSearchResultRenderer;
+            packageSearchResultRenderer = new PackageSearchResultTableRenderer(packageSearchArgs.SearchTerm, packageSearchArgs.Logger);
+            packageSearchResultRenderer.Start();
+
+            while (searchRequests.Count > 0)
+            {
+                Task<IEnumerable<IPackageSearchMetadata>> completedTask = await Task.WhenAny(searchRequests.Keys);
+                PackageSource source = searchRequests[completedTask];
+
+                IEnumerable<IPackageSearchMetadata> searchResult = null;
+
+                try
+                {
+                    searchResult = await completedTask;
+                }
+                catch (FatalProtocolException ex)
+                {
+                    // search
+                    // Throws FatalProtocolException for JSON parsing errors as fatal metadata issues.
+                    // Throws FatalProtocolException for HTTP request issues indicating critical source(v2/v3) problems.
+                    packageSearchResultRenderer.Add(source, ex.Message);
+                    searchRequests.Remove(completedTask);
+                    continue;
+                }
+                catch (OperationCanceledException ex)
+                {
+                    packageSearchResultRenderer.Add(source, ex.Message);
+                    searchRequests.Remove(completedTask);
+                    continue;
+                }
+                catch (InvalidOperationException ex)
+                {
+                    // Thrown for a local package with an invalid source destination.
+                    packageSearchResultRenderer.Add(source, ex.Message);
+                    searchRequests.Remove(completedTask);
+                    continue;
+                }
+
+                if (searchResult == null)
+                {
+                    packageSearchResultRenderer.Add(source, Strings.Error_CannotObtainSearchSource);
+                }
+                else
+                {
+                    if (packageSearchArgs.ExactMatch)
+                    {
+                        var lastResult = searchResult.LastOrDefault();
+                        if (lastResult != null)
+                        {
+                            packageSearchResultRenderer.Add(source, new[] { lastResult });
+                        }
+                    }
+                    else
+                    {
+                        packageSearchResultRenderer.Add(source, searchResult);
+                    }
+                }
+
+                searchRequests.Remove(completedTask);
+            }
+
+            packageSearchResultRenderer.Finish();
+            return 0;
```

Comment:
Say we have 2 sources, only 1 responds, but 1 times out. Do we want the exit to be 0? 
It's a genuine question. Have you checked how nuget.exe list or nuget.exe handle that?

Reply:
I hope we would still wait for and display results from other feeds in this case. For internal projects we often have many feeds and who knows whether one will have a problem for some reason..

Reply:
Yep, I agree, and it seems like the implementation is doing that as well (displaying partial results).

Just curious what should the exit code would be when we have these partial results.

Reply:
Currently the way I have it implemented is so that if a search times out for one source, we print an error, skip the source, and we continue to the next source. The only situation where we exit with 1 is if the command is badly formatted or an invalid source is specified.  If the command can be parsed and the source identified, it will always exit with 0. However, if anything goes wrong for a specific source search, the appropriate error is printed out and the search continues to the next source.

From my understanding, nuget.exe list on the other hand does not catch such exceptions, it just lets them be thrown. The nuget.exe program is designed so that if any command throws an exception, it exits with 1 and prints the error. Therefore a nuget.exe list command in the situation a source times out, will exit 1 and print the exception.

I could also have `dotnet package search` exit with 1 rather than continuing to search. I think I should make it part of the discussion in the `additional options` spec review.

Reply:
I think it's ok if we complete the operations for all sources and end up with an exit code for 1. 
A single source failing would change the result for other sources, unlike say restore for example, where a single source being inaccessible could actually change what gets resolved.